psi4/src/psi4/libfock/jk.cc

```diff
@@ -75,20 +75,38 @@ std::shared_ptr<JK> JK::build_JK(std::shared_ptr<BasisSet> primary, std::shared_
         return std::shared_ptr<JK>(jk);
 
     } else if (jk_type == "DF") {
-        DFJK* jk = new DFJK(primary, auxiliary);
-
-        if (options["INTS_TOLERANCE"].has_changed()) jk->set_cutoff(options.get_double("INTS_TOLERANCE"));
-        if (options["PRINT"].has_changed()) jk->set_print(options.get_int("PRINT"));
-        if (options["DEBUG"].has_changed()) jk->set_debug(options.get_int("DEBUG"));
-        if (options["BENCH"].has_changed()) jk->set_bench(options.get_int("BENCH"));
-        if (options["DF_INTS_IO"].has_changed()) jk->set_df_ints_io(options.get_str("DF_INTS_IO"));
-        if (options["DF_FITTING_CONDITION"].has_changed())
-            jk->set_condition(options.get_double("DF_FITTING_CONDITION"));
-        if (options["DF_INTS_NUM_THREADS"].has_changed())
-            jk->set_df_ints_num_threads(options.get_int("DF_INTS_NUM_THREADS"));
-
-        return std::shared_ptr<JK>(jk);
+        
+        if(options.get_str("DF_SCF_TYPE") == "DFJK") {
```

Comment:
We might want to consider calling the current `DFJK` to something like `DiskDFJK` or something so that `DFJK` defines a type with multiple algorithms under it. We haven't handled something like this before.

Reply:
Are you thinking another level of inheritance? Namely something like `JK` -> `DFJK` -> (`symm_DFJK` and `disk_DFJK`)?
What I mean is that we could create a generic `DFJK` derived from `JK` that would handle all the reused density fitting stuff.  Then the current `symm_JK` and `DFJK` classes would inherit from that class.  Would require some code movement.

Reply:
This comes from the fact that Im not really sure the `DF_SCF_TYPE` option is correct here. I think we should have `SymmDF` and `DiskDF` options here which call the different JK objects. We should also have a `DF` input option which chooses between `SymmDFJK` and `DiskDFJK` via something like:

```
elif (jk_type == "DF") {
    if (something){
        return build_JK(primary, auxiliary, options, "SymmJK")
   } else {
        return build_JK(primary, auxiliary, options, "DiskJK")
   }
}
```

This give users the ability to call `build_JK` with either: `SymmDF`, `DiskDF` or `DF` so they can be explicit or implicit.

Happy to take other ways of doing this, but something simple would be good here.

As a note please change the class name of `symm_JK` to `SymmJK`, classes should be in Pascal case.

Reply:
I dont know if it's at the programmer level, but certainly `SymmDF|DiskDF` is not fully standard user territory.

Reply:
So did we decide on what `something` should be?

Reply:
Lets go with `SYMM_DF` and `PACKED_DF` as `jk_type` options keys where `SymmDFJK` and `PackedDFJK` are their class names. Please note that `symm_DF` as a class name conflicts with out style choices.

Reply:
Also, I can change `SCF_TYPE` like so:
Before:
```
options.add_str("SCF_TYPE", "PK", "DIRECT DF PK OUT_OF_CORE CD GTFOCK");
```
After:
```
options.add_str("SCF_TYPE", "PK", "DIRECT PACKED_DFJK SYMM_DFJK PK OUT_OF_CORE CD GTFOCK");
```
But then how do we want to assign the `SCF_TYPE` in the pertinent locations?  The reason we went with `DF_SCF_TYPE` originally was that it allowed the reuse of most DF logic with just one difference in `jk.cc`.

Reply:
Is there ever going to be logic switching btwn `SYMM_DF` and `PACKED_DF`? If so, then `DF` (default; apply best of symm/packed), `SYMM_DF`, and `PACKED_DF` (latter two for power user control) are sensible.

If not, then `DF` (which name we mustn't change) should map internally to `SYMM_DF` and `PACKED_DF` can let power users access the orig. implementation. And the `read_options` docstring should clarify all.

Reply:
The logic is something like this:
1.  `PackedDFJK` if `wK` else `SymmDFJK`
2.  `PackedDFJK` if procedure relies intrinsically on old DFJK else `SymmDFJK`
3.  `SymmDFJK` if `SymmDFJK->AO_core()` else `PackedDFJK` 

I am not sure where this logic will be applied.  From our discussions, (3.) will probably be a conservative fully dense estimation, but we will need all of `nbf`, `naux`, and `doubles` where it is applied.  Unfortunately, `doubles` (memory allocated to JK object) is procedure dependent.

Reply:
@psi4/maintainers welcome to chime in

I think this comes down to do we collapse SCF algorithm and DFJK algorithm into one option, `SCF_TYPE` or two `SCF_TYPE` and `DF_SCF_TYPE`/`DFJK_TYPE`/? I keep being persuaded by the last person who posted, so casting a wider net for opinions.

Reply:
My take is that having the two options might be easiest.  A power user can use `DF_SCF_TYPE / DFJK_TYPE` if they want, otherwise we can map to the correct object (SymmDFJK or PackedDFJK) under the hood if `SCF_TYPE == DF`.  I guess it comes down to whoever has enough experience with Psi4 to imagine how to apply the above logic in the cleanest way.

Reply:
I agree with @schiebermc two options to allow the user who want to control it have control. possibly with a generic default like `DFJK_TYPE = DEFAULT` which would be documented to explain logic used to select the implementation by default. 

The selection logic would check the non-defualt option values, and if none of them catch the "else" implements the logic described above? 
 
I am also in favor of the `DFJK_TYPE` name for the proposed option over the `DF_SCF_TYPE` name. It is shorter and more descriptive IMO. In all cases the `SCF_TYPE` is `DF` the `DFJK` engine is what is getting selected.

Reply:
I really like this logic encoded inside the JK constructor rather than another option that I have to track down so that when I request DF it makes the correct decision for me, but I can overload it if needed. `DF_SCF_TYPE` is going to be difficult as JK is used in many places than just SCF.